src/domain/login/CompleteSSOLoginViewModel.js

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2021 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {ViewModel} from "../ViewModel.js";
+import {LoginFailure} from "../../matrix/SessionContainer.js";
+
+export class CompleteSSOLoginViewModel extends ViewModel {
+    constructor(options) {
+        super(options);
+        const {
+            loginToken,
+            sessionContainer,
+            attemptLogin,
+        } = options;
+        this._loginToken = loginToken;
+        this._sessionContainer = sessionContainer;
+        this._attemptLogin = attemptLogin;
+        this._errorMessage = "";
+        this.performSSOLoginCompletion();
+    }
+
+    get errorMessage() { return this._errorMessage; }
+
+    _showError(message) {
+        this._errorMessage = message;
+        this.emitChange("errorMessage");
+    }
+
+    async performSSOLoginCompletion() {
+        if (!this._loginToken) {
+            return;
+        }
+        const homeserver = await this.platform.settingsStorage.getString("sso_ongoing_login_homeserver");
+        const loginOptions = await this._sessionContainer.queryLogin(homeserver);
```

Comment:
Hmm, this should probably have a try catch around it catching any error and calling showError